File: app/src/main/java/com/wyang/study/ui/util/TreeHelper.java
package com.wyang.study.ui.util;

import com.wyang.study.utils.NodeSeeker;
import com.wyang.study.utils.TreeNode;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by weiyang on 2019-09-22.
 */
public class TreeHelper {
    private TreeNode rootNode;
    private Map<String, TreeNode> areaMap;//区的map,提升解析村级别效率过慢问题

    public TreeHelper() {
        rootNode = new TreeNode(null);
        areaMap = new HashMap<>();
    }

    public void putAreaNode(TreeNode node) {
        areaMap.put(node.getAttribute("code"), node);
    }

    public TreeNode getAreaNode(String code) {
        return areaMap.get(code);
    }

    public NodeSeeker rootSeeker() {
        return NodeSeeker.newInstance(rootNode);
    }

    public NodeSeeker provincesSeeker() {
        return rootSeeker().children();
    }

    public NodeSeeker areaSeeker() {
        return provincesSeeker().children();
    }

    public NodeSeeker citiesSeeker() {
        return areaSeeker().children();
    }

    public NodeSeeker streetsSeeker() {
        return citiesSeeker().children();
    }

    public TreeNode getProvince(String code) {
        return rootSeeker()
                .children()
                .attribute("code", code)
                .firstResult();
    }

    public TreeNode getCity(String code) {
        return provincesSeeker()
                .children()
                .attribute("code", code)
                .firstResult();
    }

    public TreeNode getArea(String code) {
        return citiesSeeker()
                .children()
                .attribute("code", code)
                .firstResult();
    }

    public TreeNode getStreet(String code) {
        return areaSeeker()
                .children()
                .attribute("code", code)
                .firstResult();
    }

}

File: app/src/main/java/com/wyang/study/ui/util/TreeHelper.kt
package com.wyang.study.ui.util

import android.text.TextUtils
import com.wyang.study.NodeSeeker
import com.wyang.study.TreeNode

class TreeHelper {
    private val rootNode by lazy {
        TreeNode(null)
    }
    private val areaMap by lazy {
        HashMap<String, TreeNode>()
    }

    fun putAreaNode(node: TreeNode) {
        areaMap[node["code"] as String] = node
    }

    fun getAreaNode(code: String?): TreeNode? {
        return areaMap[code]
    }

    fun rootSeeker(): NodeSeeker {
        return NodeSeeker.newInstance(rootNode)
    }

    fun provincesSeeker(): NodeSeeker {
        return rootSeeker().children()
    }

    private fun areaSeeker(): NodeSeeker {
        return provincesSeeker().children()
    }

    private fun citiesSeeker(): NodeSeeker {
        return areaSeeker().children()
    }

    fun streetsSeeker(): NodeSeeker {
        return citiesSeeker().children()
    }

    fun getProvince(code: String?): TreeNode {
        return rootSeeker()
            .children()
            .match {
                return@match TextUtils.equals(it["code"] as String, code)
            }
            .firstResult()
    }

    fun getCity(code: String?): TreeNode {
        return provincesSeeker()
            .children()
            .match {
                return@match TextUtils.equals(it["code"] as String, code)
            }
            .firstResult()
    }

    fun getArea(code: String?): TreeNode {
        return citiesSeeker()
            .children()
            .match {
                return@match TextUtils.equals(it["code"] as String, code)
            }
            .firstResult()
    }

    fun getStreet(code: String?): TreeNode {
        return areaSeeker()
            .children()
            .match {
                return@match TextUtils.equals(it["code"] as String, code)
            }
            .firstResult()
    }
}